File: src/main/java/com/jcloisterzone/game/capability/TunnelCapability.java
package com.jcloisterzone.game.capability;

import com.jcloisterzone.Player;
import com.jcloisterzone.action.TunnelAction;
import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.game.Capability;
import com.jcloisterzone.game.Rule;
import com.jcloisterzone.game.Token;
import com.jcloisterzone.game.state.ActionsState;
import com.jcloisterzone.game.state.Flag;
import com.jcloisterzone.game.state.GameState;
import com.jcloisterzone.game.state.PlacedTunnelToken;
import com.jcloisterzone.random.RandomGenerator;
import io.vavr.Predicates;
import io.vavr.Tuple2;
import io.vavr.collection.HashMap;
import io.vavr.collection.List;
import io.vavr.collection.Map;
import io.vavr.collection.Set;

import java.util.ArrayList;

/**
 * Capability model is {@code Map<FeaturePointer, String>} - tunnels,
 *   key is tunnel token id or null if nothing is placed
 */
public final class TunnelCapability extends Capability<Map<FeaturePointer, PlacedTunnelToken>> {

	public enum Tunnel implements Token {
		TUNNEL_A,
		TUNNEL_B,
		TUNNEL_C
    }

	private static final long serialVersionUID = 1L;

    @Override
    public GameState onStartGame(GameState state, RandomGenerator random) {
        int playersCount = state.getPlayers().getPlayers().length();
        String moreTokens = state.getStringRule(Rule.MORE_TUNNEL_TOKENS);
        state = state.mapPlayers(ps -> {
            ps = ps.setTokenCountForAllPlayers(Tunnel.TUNNEL_A, 2);
            if (playersCount == 3 && "3/2".equals(moreTokens)) {
                ps = ps.setTokenCountForAllPlayers(Tunnel.TUNNEL_B, 2);
            }
            if (playersCount < 3) {
                if (!"1/1".equals(moreTokens)) {
                    ps = ps.setTokenCountForAllPlayers(Tunnel.TUNNEL_B, 2);
                    if ("3/2".equals(moreTokens)) {
                        ps = ps.setTokenCountForAllPlayers(Tunnel.TUNNEL_C, 2);
                    }
                }
            }
            return ps;
        });
        state = setModel(state, HashMap.empty());
        return state;
    }

    @Override
    public GameState onActionPhaseEntered(GameState state) {
        Player player = state.getPlayerActions().getPlayer();

        if (state.hasFlag(Flag.TUNNEL_PLACED)) {
            return state;
        }

        java.util.List<TunnelAction> actions = createTunnelActions(state);
        if (actions.isEmpty()) {
            return state;
        }

        ActionsState as = state.getPlayerActions();
        for (TunnelAction action: actions) {
            as = as.appendAction(action);
        }
        return state.setPlayerActions(as);
    }

    public java.util.List<TunnelAction> createTunnelActions(GameState state) {
        java.util.List actions = new ArrayList(3);
        Player player = state.getTurnPlayer();

        Set<FeaturePointer> openTunnels = getModel(state)
                .filterValues(Predicates.isNull())
                .map(Tuple2::_1)
                .toSet();

        if (openTunnels.isEmpty()) {
            return actions;
        }

        ActionsState as = state.getPlayerActions();
        for (Tunnel token : Tunnel.values()) {
            if (state.getPlayers().getPlayerTokenCount(player.getIndex(), token) == 0) {
                continue;
            }
            actions.add(new TunnelAction(openTunnels, token));
        }
        return actions;
    }
}

File: src/main/ts/com/jcloisterzone/game/capability/TunnelCapability.ts
import { HashMap, type Map as VMap } from "../../../../io/vavr/Map.js";
import { HashSet, type Set } from "../../../../io/vavr/Set.js";
import { Vector } from "../../../../io/vavr/SeqTypes.js";
import { JavaEnum } from "../../../../lang/JavaEnum.js";
import type { FeaturePointer } from "../../board/pointer/FeaturePointer.js";
import { TunnelAction } from "../../action/TunnelAction.js";
import type { PlayerAction } from "../../action/PlayerAction.js";
import type { RandomGenerator } from "../../random/RandomGenerator.js";
import { PlaceTunnel } from "../../reducers/PlaceTunnel.js";
import { Capability } from "../Capability.js";
import { Rule } from "../Rule.js";
import type { Token } from "../Token.js";
import { ActionsState } from "../state/ActionsState.js";
import { Flag } from "../state/Flag.js";
import type { GameState } from "../state/GameState.js";
import type { PlacedTunnelToken } from "../state/PlacedTunnelToken.js";

import { isInstanceOfFeatureCompletionBlocker } from "./trait/FeatureCompletionBlocker.js";

/** Tunnels (mini-expansion). Model: map of placed tunnel tokens. */
export class TunnelCapability extends Capability<VMap<FeaturePointer, PlacedTunnelToken>> {
  override onStartGame(state: GameState, random: RandomGenerator): GameState {
    const playersCount = state.getPlayers().getPlayers().length();
    const moreTokens = state.getStringRule(Rule.MORE_TUNNEL_TOKENS);
    const T = TunnelCapability.Tunnel;
    state = state.mapPlayers((ps) => {
      ps = ps.setTokenCountForAllPlayers(T.TUNNEL_A, 2);
      if (playersCount === 3 && moreTokens === "3/2") {
        ps = ps.setTokenCountForAllPlayers(T.TUNNEL_B, 2);
      }
      if (playersCount < 3) {
        if (moreTokens !== "1/1") {
          ps = ps.setTokenCountForAllPlayers(T.TUNNEL_B, 2);
          if (moreTokens === "3/2") {
            ps = ps.setTokenCountForAllPlayers(T.TUNNEL_C, 2);
          }
        }
      }
      return ps;
    });
    return this.setModel(state, HashMap.empty<FeaturePointer, PlacedTunnelToken>());
  }

  override onActionPhaseEntered(state: GameState): GameState {
    if (state.hasFlag(Flag.TUNNEL_PLACED)) return state;
    const actions = this.createTunnelActions(state);
    if (actions.length === 0) return state;
    let as = state.getPlayerActions()!;
    for (const action of actions) as = as.appendAction(action as unknown as PlayerAction<unknown>);
    return state.setPlayerActions(as);
  }

  /** A TunnelAction per tunnel-token colour the turn player still holds, over the open
   *  tunnel ends. With the last token of a colour, ends whose completion would be blocked
   *  (e.g. by a donkey) are excluded. */
  createTunnelActions(state: GameState): TunnelAction[] {
    const model = this.getModel(state);
    const openTunnels: Set<FeaturePointer> = HashSet.ofAll(
      model.filterValues((v) => v === null).toArray().map((t) => t._1),
    );
    if (openTunnels.isEmpty()) return [];

    const player = state.getTurnPlayer()!;
    const activePlayer = state.getActivePlayer();
    const actions: TunnelAction[] = [];
    for (const token of TunnelCapability.Tunnel.values()) {
      const count = state.getPlayers().getPlayerTokenCount(player.getIndex(), token);
      if (count === 0) continue;
      let allowedTunnels: Set<FeaturePointer> = openTunnels;
      if (count === 1) {
        allowedTunnels = HashSet.empty<FeaturePointer>();
        for (const cap of state.getCapabilities().toSeq()) {
          if (!isInstanceOfFeatureCompletionBlocker(cap)) continue;
          for (const fp of openTunnels) {
            const as =
              activePlayer !== null
                ? state.getPlayerActions()!.appendAction(new TunnelAction(openTunnels, token) as unknown as PlayerAction<unknown>)
                : new ActionsState(player, Vector.of(new TunnelAction(openTunnels, token) as unknown as PlayerAction<unknown>), true);
            let sim = state.setPlayerActions(as);
            sim = new PlaceTunnel(token, fp).apply(sim);
            if (!cap.isFeatureCompletionBlocked(sim, fp)) allowedTunnels = allowedTunnels.add(fp);
          }
        }
      }
      actions.push(new TunnelAction(allowedTunnels, token));
    }
    return actions;
  }
}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace TunnelCapability {
  export class Tunnel extends JavaEnum implements Token {
    static readonly TUNNEL_A = new Tunnel("TUNNEL_A", 0);
    static readonly TUNNEL_B = new Tunnel("TUNNEL_B", 1);
    static readonly TUNNEL_C = new Tunnel("TUNNEL_C", 2);
    private static readonly VALUES: readonly Tunnel[] = [Tunnel.TUNNEL_A, Tunnel.TUNNEL_B, Tunnel.TUNNEL_C];
    static values(): readonly Tunnel[] {
      return Tunnel.VALUES;
    }
    static valueOf(name: string): Tunnel {
      const v = Tunnel.VALUES.find((t) => t.name() === name);
      if (v === undefined) throw new Error("No tunnel token " + name);
      return v;
    }
  }
}

Capability.register(TunnelCapability);
